### PR TITLE
Fixed autobonus on the same equip slot

### PR DESCRIPTION
**Pull request by Lemongrass3110**
* **Addressed Issue(s)**: #2079

* **Server Mode**: Both

* **Description of Pull Request**: 
If two different autobonuses would be triggered from the same equipment slot, only one of them would work.
With this now all autobonuses of a slot will work.

Thanks to @LunarSHINING for reporting and testing
Thanks to @Tokeiburu for the initial report

### CODE REVIEW

File: src/map/pc.cpp [hunk 1]
```diff
@@ -10785,6 +10804,36 @@ bool pc_equipitem(struct map_session_data *sd,short n,int req_pos,bool equipswit
 	return true;
 }
 
+void pc_deleteautobonus( struct map_session_data& sd, std::vector<s_autobonus>& bonus, int position ){
```
Comment (Atemo) [inline src/map/pc.cpp:10807]: `struct map_session_data& sd` could be removed since sd is not used in the function?

File: src/map/pc.cpp [hunk 2]
```diff
@@ -2772,8 +2772,17 @@ static void pc_bonus_item_drop(std::vector<s_add_drop> &drop, t_itemid nameid, u
  * @param onskill: Skill used to trigger autobonus
  * @return True on success or false otherwise
  */
-bool pc_addautobonus(std::vector<s_autobonus> &bonus, const char *script, short rate, unsigned int dur, uint16 flag, const char *other_script, unsigned int pos, bool onskill)
-{
+bool pc_addautobonus(std::vector<s_autobonus> &bonus, const char *script, short rate, unsigned int dur, uint16 flag, const char *other_script, unsigned int pos, bool onskill){
+	// Check if the same bonus already exists
+	for( std::vector<s_autobonus>::iterator it = bonus.begin(); it != bonus.end(); it++ ){
```
Comment (aleos89) [inline src/map/pc.cpp:2777]: Adjust this to a normal iterator for easier reading!

File: src/map/pc.cpp [hunk 3]
```diff
@@ -2836,23 +2845,35 @@ void pc_delautobonus(struct map_session_data* sd, std::vector<s_autobonus> &bonu
...
 					}
 
-					if ((equip_pos_idx&b.pos) == b.pos)
+					if( ( equip_pos_idx&b.pos ) == b.pos ){
 						script_run_autobonus(b.bonus_script, sd, b.pos);
+					}else{
+						// Not all required items equipped anymore
+						restore = false;
+						delete_timer( b.active, pc_endautobonus );
+						b.active = INVALID_TIMER;
```
Comment (aleos89) [inline src/map/pc.cpp:2854]: Move to a destructor in s_autobonus.

File: src/map/pc.cpp [hunk 4]
```diff
@@ -2836,23 +2845,35 @@ void pc_delautobonus(struct map_session_data* sd, std::vector<s_autobonus> &bonu
...
-		if (b.bonus_script)
+		if( restore ){
+			it++;
+			continue;
+		}
+
+		if( b.bonus_script ){
 			aFree(b.bonus_script);
-		if (b.other_script)
+			b.bonus_script = nullptr;
```
Comment (aleos89) [inline src/map/pc.cpp:2870]: Move to a destructor in s_autobonus.

File: src/map/pc.cpp [hunk 5]
```diff
@@ -2836,23 +2845,35 @@ void pc_delautobonus(struct map_session_data* sd, std::vector<s_autobonus> &bonu
...
+		}
+
+		if( b.other_script ){
 			aFree(b.other_script);
+			b.other_script = nullptr;
```
Comment (aleos89) [inline src/map/pc.cpp:2875]: Move to a destructor in s_autobonus.

File: src/map/pc.cpp [hunk 6]
```diff
@@ -10785,6 +10804,36 @@ bool pc_equipitem(struct map_session_data *sd,short n,int req_pos,bool equipswit
...
+		s_autobonus b = *it;
+
+		if( ( b.pos & position ) != b.pos ){
+			it++;
+			continue;
+		}
+
+		if( b.active != INVALID_TIMER ){
+			delete_timer( b.active, pc_endautobonus );
+			b.active = INVALID_TIMER;
```
Comment (aleos89) [inline src/map/pc.cpp:10820]: Move to a destructor in s_autobonus.

File: src/map/pc.cpp [hunk 7]
```diff
@@ -10785,6 +10804,36 @@ bool pc_equipitem(struct map_session_data *sd,short n,int req_pos,bool equipswit
...
+		}
+
+		if( b.bonus_script ){
+			aFree( b.bonus_script );
+			b.bonus_script = nullptr;
```
Comment (aleos89) [inline src/map/pc.cpp:10825]: Move to a destructor in s_autobonus.

File: src/map/pc.cpp [hunk 8]
```diff
@@ -10785,6 +10804,36 @@ bool pc_equipitem(struct map_session_data *sd,short n,int req_pos,bool equipswit
...
+		}
+
+		if( b.other_script ){
+			aFree( b.other_script );
+			b.other_script = nullptr;
```
Comment (aleos89) [inline src/map/pc.cpp:10830]: Move to a destructor in s_autobonus.

**Review by aleos89 — APPROVED** [review]
Just a few optimizations! Rest looks good!

**Pull request merged**